### PR TITLE
add lower bound slot to utxo by address query

### PR DESCRIPTION
**Pull request by kayvank**
PLT-5566, add lower bound slotNo to utxo queries

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting main unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [ ] Reviewer requested

### CODE REVIEW

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 1]
```diff
@@ -134,6 +136,8 @@ tests = testGroup "Spec.Marconi.ChainIndex.Indexers.Utxo"
         ]
     ]
 
+deriving instance Ord (StorableResult Utxo.UtxoHandle)
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:139]: That should go in the UtxoIndexer module.

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 2]
```diff
@@ -3,9 +3,11 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE RecordWildCards   #-}
 
+{-# OPTIONS_GHC -Wno-orphans #-}
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:6]: Bin.

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 3]
```diff
@@ -314,41 +319,76 @@ propSaveAndRetrieveUtxoEvents = property $ do
...
+        forAll $ genEventWithShelleyAddressAtChainPoint cp
+      indexer <- liftIO $ raiseException $
+        Utxo.open ":memory:" (Utxo.Depth 1) False >>= Storable.insertMany events
+      let
+        addresses :: [C.AddressAny]
+        addresses = concatMap (\e -> Utxo.ueUtxos e ^.. folded . Utxo.address) events
+
+      upperBoundSlotNo@(C.SlotNo upperBoundSlotNoWord) <- getSlot cp
+
+    -- pick a slotNo less than highPoint and more than genesis
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:345]: Indent.

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 4]
```diff
@@ -314,41 +319,76 @@ propSaveAndRetrieveUtxoEvents = property $ do
...
-             >>= Storable.insertMany events
-  let _slot = getSlot $ chainPoints !! (length chainPoints `div` 2)
-      qAddresses
-        = List.nub  -- remove duplicate addresses
-        . fmap (flip Utxo.UtxoByAddress _slot . Utxo._address)
-        . concatMap (Set.toList . Utxo.ueUtxos)
-        $ events
-  results <- liftIO . raiseException . traverse (Storable.query Storable.QEverything indexer) $ qAddresses
-  let filterResult = \case
+--   - retrieve for all addresses starting from genesis, is the same as retrieving for all address prior to last observed SlotNo
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:327]: I think there's syntax issues with this sentence.

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 5]
```diff
@@ -314,41 +319,76 @@ propSaveAndRetrieveUtxoEvents = property $ do
...
+--   - retrieve all addresses in a given open interval should not have any slotNo outside of the query interval
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:328]: retrieve -> retrieving

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 6]
```diff
@@ -543,6 +596,18 @@ propLastChainPointOnRewindedIndexer = property $ do
     result <- liftIO $ raiseException $ Storable.query Storable.QEverything indexer'' LastSyncPoint
     result === LastSyncPointResult lastestPointPostRollback
 
-getSlot :: C.ChainPoint -> Maybe C.SlotNo
-getSlot C.ChainPointAtGenesis = Nothing
-getSlot (C.ChainPoint s _)    = Just s
+-- | make unique queries
+mkUtxoQueries
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:600]: I think this function is not useful outside of propUtxoQueryByAddressAndSlotInterval, right? Should it go inside a `where` clause?

File: marconi-sidechain/src/Marconi/Sidechain/Api/Routes.hs [hunk 7]
```diff
@@ -111,20 +111,25 @@ instance FromJSON GetCurrentSyncedBlockResult where
 
 data GetUtxosFromAddressParams
     = GetUtxosFromAddressParams
-    { queryAddress       :: !String
-    , queryUnspentAtSlot :: !(Maybe Word64)
-    } deriving (Eq, Show)
+    { queryAddress             :: !String -- ^ address to query for
+    , queryCreatedAfterSlotNo  :: !Word64 -- ^ lower bound slotNo to query for
+    , queryUnspentBeforeSlotNo :: !(Maybe Word64) -- ^ optional upper bound slotNo to query for
```
Comment (koslambrou) [inline marconi-sidechain/src/Marconi/Sidechain/Api/Routes.hs:116]: It's the other way around. `queryUnspentBeforeSlotNo` is mandatory and the other one is optional given the `API.adoc`

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs [hunk 8]
```diff
@@ -90,6 +91,52 @@ import Marconi.Core.Storable qualified as Storable
...
+          <> " is not less than or equal to upperBound "
+          <> show y
+
+  in wrap BothOpen p p'
+
+isInInterval :: Interval C.SlotNo -> C.ChainPoint -> Bool
+isInInterval slotNoInterval = \case
+  C.ChainPointAtGenesis -> slotNoInterval == Empty
+  (C.ChainPoint slotNo _)  -> case slotNoInterval of
+    Empty                     -> True -- Genesis
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs:135]: "Genesis"?

**Review by berewt — COMMENTED** [review]
Not sure about the interval representation and interest. The rest look good to me.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs [hunk 9]
```diff
@@ -98,10 +145,16 @@ data UtxoHandle = UtxoHandle
...
 
+data QueryUtxoByAddress where
+  QueryUtxoByAddress :: !C.AddressAny
+                        -> !(Interval C.SlotNo)
+                        -> QueryUtxoByAddress
+  deriving (Show, Eq)
+
 data instance StorableQuery UtxoHandle
-    = UtxoByAddress C.AddressAny (Maybe C.SlotNo)
+    = QueryWrapper QueryUtxoByAddress
```
[the remark this answers is not in the document]
Comment (kayvank) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs:155]: Using `StorableQuery UtxoHandle` will force
- use pattern matching in all function that supports the query by address. 
- or alternatively pass the arguments individually  to those functions  
Examples are: findByAdddres, queryBuffer, eventsAtAddress

As for the GADT, that was the linter suggestion, which I didn't have to comply with, but didnt see any harm.

[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs:155]: Not sure I follow here `data instance StorableQuery UtxoHandle = QueryUtxoByAddress !C.AddressAny !(Interval C.SlotNo)` looks isomorphic to what we have currently and just remove one indirection. Am I missing something?

[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs:155]: Oh never mind, I somehow miss the other part of the sum types.

[the remark this answers is not in the document]
Comment (kayvank) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs:155]: @berewt after our review today, I changed from GADT to Jut a good old data type, 
```data QueryUtxoByAddress = QueryUtxoByAddress !C.AddressAny !(Interval C.SlotNo)
```

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs [hunk 10]
```diff
@@ -90,6 +91,52 @@ import Marconi.Core.Storable qualified as Storable
...
+
+interval
+  :: (Ord r, Show r)
+  => Maybe r -- ^ lower bound
+  -> Maybe r  -- ^ upper bound
+  -> Either IndexerError (Interval r)
+interval Nothing Nothing  = Right intervalFromGenesis
+interval Nothing (Just p) = Right $ LessThanOrEqual p
+interval (Just p) Nothing = Right $ GreaterOrEqual p
+interval (Just p) (Just p') =
```
[the remark this answers is not in the document]
Comment (kayvank) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs]: this is simiplified now based [getUxFromAddress Request Schema](https://github.com/input-output-hk/marconi/blob/main/marconi-sidechain/doc/API.adoc#json-schema-1)

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs [hunk 11]
```diff
@@ -341,13 +387,13 @@ open dbPath (Depth k) isToVacuume = do
                       , value BLOB
                       , inlineScript BLOB
                       , inlineScriptHash BLOB
-                      , slotNo INT NOT NULL
+                      , slotNo NUMERIC NOT NULL
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs:390]: Isn't numeric worse than int? (as it subject to precision lost)

**Review by berewt — COMMENTED** [review]
LGTM.

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 12]
```diff
@@ -312,141 +326,157 @@ propSaveAndRetrieveUtxoEvents = property $ do
...
---   - Insert many events at various chainPoints
---   - Fetch for all the evenent addresses
---   - There should not be any results with ChainPoint > highChainPoint
-propUtxoQueryByAddressAndSlot :: Property
-propUtxoQueryByAddressAndSlot = property $ do
+ ---------------------------------------------------------------------
+-- TODO --
+-- Uncommenting the TODO section below will fail the test, see ticket PLT-5937
+-- Add this to observe the Integer Overflow in the logs:
+-- liftIO $ SQL.setTrace (Utxo.hdlConnection (h ^. Storable.handle)) $ Just (Text.IO.appendFile "utxoSQLiteTrace.loger")
```
Comment (kayvank) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:340]: @berewt  adding this line to right after the storage `open`  sill log the SQL statements.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs (hunk 11 above)
Comment (kayvank) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs]: No, it made no difference in my tests.  From the docs, it gets converted to REAL, but still a 64 bit signed.  See [SQLite docs, section 3, Type Affinity](https://www.sqlite.org/datatype3.html)
I also tested it and examined the output using sqlite [setTrace](https://hackage.haskell.org/package/sqlite-simple-0.4.18.2/docs/Database-SQLite-Simple.html#v:setTrace) and still got `-1` for Word64.maxBond.  The relevant code is in the utxoIndexer.hs & commented with TODO.

I have changed back to INT.

So in summary, the fix is store SlotNo as either:
- Text
- Binary
Either way is a trivial change as the  Cardano API has the corresponding instances for SlotNo

Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs]: Yes, but real is a iee float, that is subject to precision loss on high value (the -1 on `maxBound` is puzzling in this case). I think a signed integer is better than a real at the moment, as we should be fine with the 63bits of positive integers for quite a long time). No need to add the confusion of numeric that is a bit further than integer to represent the initial data.

**Review by berewt — APPROVED** [review]
Looks good to me.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs (hunk 9 above)
[the remark this answers is not in the document]
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs:146]: That comment doesn't seem resolved.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs [hunk 13]
```diff
@@ -435,10 +479,10 @@ instance Buffered UtxoHandle where
               slotNo
            ORDER BY
               slotNo DESC
-           LIMIT ?|] (SQL.Only sz) :: IO [[Integer]]
+           LIMIT ?|] (SQL.Only sz) :: IO [[Word64]]
 
     -- Take the slot number of the sz'th slot
-    let sn = if null sns
+    let sn::Word64 = if null sns
```
Comment (koslambrou) [inline marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs:485]: Bin `Word64`. It's infered afaik

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs (hunk 9 above)
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs:146]: I discussed it with @kayvank and we may need clarification: the UtxoIndexer has 2 queries, one for utxos and one for timeslot, so if we don't use an intermediary constructor and put everything in `QueryWrapper`, we either need to pass each of its field as parameters to the functions that currently take a `QueryUtxoByAddress` or pass a `StorableQuery UtxoHandle` and deal with the `LastSyncPoint` query (and thus introduce an error path).
It seems to both of us that the intermediary type was the best option.

[the remark this answers is not in the document]
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs:146]: Ahh I get it now. But can we change the name to `QueryUtxoByAddressWrapper` or something more significant? I know it's long, but `QueryWrapper` is too generic.
